File: apps/ios/Packages/TrackEvolutionKit/Tests/TrackEvolutionKitTests/VideoFixtures.swift
import Foundation
import Testing

@testable import TrackEvolutionKit

/// Access to `contracts/logic/video/` and `contracts/logic/video-parsers.json` —
/// the committed synthetic MP4s and what the JS parsers made of them.
///
/// Read from the repo rather than bundled as package resources, for the reason
/// `Goldens` gives: a copy would go stale exactly when it mattered. Regenerate
/// both with `npm run contracts:logic`.
enum VideoFixtures {
    static let directory = RepoRoot.path("contracts/logic/video")

    /// The bytes of one fixture clip, as a parser sees them.
    static func source(_ file: String) throws -> DataByteSource {
        DataByteSource(try Data(contentsOf: directory.appending(path: file)))
    }

    static func parse(_ file: String) throws -> ParsedTelemetry {
        try Telemetry.parseTelemetryFile(try source(file))
    }

    // MARK: - The pinned JS output

    static let fixture: Fixture = {
        let data = try! Data(contentsOf: RepoRoot.path("contracts/logic/video-parsers.json"))
        return try! JSONDecoder().decode(Fixture.self, from: data)
    }()

    static func expected(_ file: String) -> Case {
        guard let entry = fixture.files.first(where: { $0.file == file }) else {
            fatalError("no video fixture named \(file) — run `npm run contracts:logic`")
        }
        return entry
    }

    struct Fixture: Decodable {
        var description: String
        var gpsStride: Int
        var files: [Case]
        var afterBatchAnchor: [Anchored]
    }

    struct Case: Decodable {
        var file: String
        var note: String
        var expected: Parsed
        var picked: Picked?
    }

    struct Anchored: Decodable {
        var file: String
        var expected: Parsed
    }

    struct Parsed: Decodable {
        var kind: String
        var date: String?
        var time: String?
        var durationS: Double
        var needsLine: Bool
        var beaconCount: Int
        var metrics: Metrics?
        var gpsCount: Int
        var gpsSample: [GPSSample]?
        var laps: [Lap]
        var lapChannels: SessionChannels?
        var lapRecovery: Recovery?
        var channels: RawChannels?
    }

    struct Metrics: Decodable {
        var topSpeedKph: Double?
        var maxRpm: Double?
        var maxLatG: Double?
        var maxBrakeG: Double?
        var maxBoostKpa: Double?
        var maxOilC: Double?
    }

    struct GPSSample: Decodable {
        var i: Int
        var t: Double
        var lat: Double
        var lon: Double
        var v: Double?
    }

    struct Lap: Decodable {
        var lapNumber: Int?
        var timeMs: Int
        var estimated: Bool
        var startT: Double?
        var endT: Double?
    }

    struct Recovery: Decodable {
        var lapM: Double
        var r: Double
        var anchored: Bool
        var phaseR: Double?
        var lapCount: Int
    }

    struct RawChannels: Decodable {
        var latCount: Int
        var odoCount: Int
        var odoFirst: ChannelPoint?
        var odoLast: ChannelPoint?
    }

    struct Picked: Decodable {
        var pickedIndex: Int
        var gate: Geo.Gate
        var laps: [Lap]
        var bestLapTrace: [TracePoint]?
        var lapChannels: SessionChannels?
    }
}

File: apps/android/core/src/test/kotlin/app/trackevolution/core/telemetry/VideoFixtures.kt
package app.trackevolution.core.telemetry

import app.trackevolution.core.RepoRoot
import app.trackevolution.core.model.SessionChannels
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Access to `contracts/logic/video/` and `contracts/logic/video-parsers.json` —
 * the committed synthetic MP4s and what the JS parsers made of them.
 *
 * Read from the repo rather than copied under `src/test/resources`, for the
 * reason `Goldens` gives: a copy would go stale exactly when it mattered.
 * Regenerate both with `npm run contracts:logic`.
 */
object VideoFixtures {
    private val directory = RepoRoot.path("contracts/logic/video")

    /** The bytes of one fixture clip, as a parser sees them. */
    fun source(file: String): ByteArraySource = ByteArraySource(directory.resolve(file).readBytes())

    fun parse(file: String): ParsedTelemetry = Telemetry.parseTelemetryFile(source(file))

    // ---- The pinned JS output ------------------------------------------------

    private val lenient = Json { ignoreUnknownKeys = true }

    val fixture: Fixture by lazy {
        lenient.decodeFromString(Fixture.serializer(), RepoRoot.path("contracts/logic/video-parsers.json").readText())
    }

    fun expected(file: String): Case =
        fixture.files.firstOrNull { it.file == file }
            ?: error("no video fixture named $file — run `npm run contracts:logic`")

    @Serializable
    data class Fixture(
        val description: String,
        val gpsStride: Int,
        val files: List<Case>,
        val afterBatchAnchor: List<Anchored>,
    )

    @Serializable
    data class Case(val file: String, val note: String, val expected: Parsed, val picked: Picked? = null) {
        // Shown by JUnit as the parameterized case name.
        override fun toString(): String = file
    }

    @Serializable
    data class Anchored(val file: String, val expected: Parsed)

    @Serializable
    data class Parsed(
        val kind: String,
        val date: String? = null,
        val time: String? = null,
        val durationS: Double,
        val needsLine: Boolean,
        val beaconCount: Int,
        val metrics: Metrics? = null,
        val gpsCount: Int,
        val gpsSample: List<GpsSample>? = null,
        val laps: List<Lap>,
        val lapChannels: SessionChannels? = null,
        val lapRecovery: Recovery? = null,
        val channels: RawChannels? = null,
    )

    @Serializable
    data class Metrics(
        val topSpeedKph: Double? = null,
        val maxRpm: Double? = null,
        val maxLatG: Double? = null,
        val maxBrakeG: Double? = null,
        val maxBoostKpa: Double? = null,
        val maxOilC: Double? = null,
    )

    @Serializable
    data class GpsSample(val i: Int, val t: Double, val lat: Double, val lon: Double, val v: Double? = null)

    @Serializable
    data class Lap(
        val lapNumber: Int? = null,
        val timeMs: Int,
        val estimated: Boolean,
        val startT: Double? = null,
        val endT: Double? = null,
    )

    @Serializable
    data class Recovery(val lapM: Double, val r: Double, val anchored: Boolean, val phaseR: Double? = null, val lapCount: Int)

    @Serializable
    data class Pt(val t: Double, val v: Double)

    @Serializable
    data class RawChannels(val latCount: Int, val odoCount: Int, val odoFirst: Pt? = null, val odoLast: Pt? = null)

    @Serializable
    data class GateJson(
        val x: Double,
        val y: Double,
        val hx: Double? = null,
        val hy: Double? = null,
        val x1: Double,
        val y1: Double,
        val x2: Double,
        val y2: Double,
    )

    @Serializable
    data class Picked(
        val pickedIndex: Int,
        val gate: GateJson,
        val laps: List<Lap>,
        /** `[x, y, v]` triples. */
        val bestLapTrace: List<List<Double>>? = null,
        val lapChannels: SessionChannels? = null,
    )
}

/**
 * `#require` for JUnit: fail the test unless [value] is present, and hand it back
 * non-null. `Assertions.assertNotNull` returns void, which is why this exists.
 */
fun <T : Any> given(value: T?, label: String = "expected a value"): T =
    value ?: throw AssertionError(label)
